### PR TITLE
Make redis dependency optional

### PR DESCRIPTION
**Pull request by ahesford**
The redis package is only used when somebody uses the redis cache handler, so the imports can be restructured to make redis an optional dependency.

### CODE REVIEW

**Comment by ahesford** [comment]
Ping

**Comment by Peter-Schorn** [comment]
We could catch the import error and display a detailed error message indicating that redis is an optional dependency and must be explicitly installed (clients often expect pip install -r requirements.txt to install all the dependencies for a package).

**Comment by ahesford** [comment]
For starters, I added a note to the readme in the latest commit.

**Comment by stephanebruckert** [comment]
> We could catch the import error and display a detailed error message indicating that redis is an optional dependency and must be explicitly installed (clients often expect pip install -r requirements.txt to install all the dependencies for a package).

True, exceptionally for this use case of "connectors" we should go with the warning. @ahesford this is how a warning can be raised https://github.com/plamere/spotipy/blob/fa44fed76d7aecb70f9ee6cde14da5e2ce1e17d7/spotipy/client.py#L1951

> For starters, I added a note to the readme in the latest commit.

Thanks that helps. Note that in any case we won't be able to merge this into v2 aka master just now, because it would break existing apps. So this will have to go into v3 https://github.com/plamere/spotipy/tree/v3
See https://github.com/plamere/spotipy/issues/652
Can you please change the base of your PR from `[plamere:master](https://github.com/plamere/spotipy)` to `[plamere:v3](https://github.com/plamere/spotipy/tree/v3)`? Feel free to create a new PR if it gets messy

**Comment by ahesford** [comment]
Rebased on v3. That branch had no redis cache handler at all, so I brought the whole class over.

I added an import check during creation of a cache object to make sure redis is available and warn if not. I'm not sure there would be a good way to handle this in the functions where the import are actually needed, unless the exception catching is moved from `RedisError` to `Exception` to issue a blanket warning instead of hard errors for unexpected, non-redis failures.

File: spotipy/cache_handler.py [hunk 1]
```diff
@@ -107,3 +108,45 @@ def get_cached_token(self):
...
+            * key: May be supplied, will otherwise be generated
+                   (takes precedence over `token_info`)
+        """
+        self.redis = redis
+        self.key = key if key else 'token_info'
+
+        try:
+            from redis import RedisError
+        except ImportError:
+            warnings.warn('Error importing requied module "redis"', UserWarning)
```
Comment (Peter-Schorn) [inline spotipy/cache_handler.py:132]: There's a typo here: "requied".

Comment (Peter-Schorn) [reply spotipy/cache_handler.py:132]: Also, the word "required" may imply to some users that this module is required for the whole package. Instead, the error message should indicate that this module is required *for the `RedisCacheHandler`*.

Comment (ahesford) [reply spotipy/cache_handler.py]: Fixed. I also restructured the history to preserve the evolution of Redis support on the master branch and diminish the extent of my own changes.

File: spotipy/cache_handler.py [hunk 2]
```diff
@@ -107,3 +113,49 @@ def get_cached_token(self):
...
+        """
+        self.redis = redis
+        self.key = key if key else 'token_info'
+
+        try:
+            from redis import RedisError
+        except ImportError:
+            warnings.warn(
+                'Error importing module "redis", '
+                'required for RedisCacheHandler',
```
Comment (Peter-Schorn) [inline spotipy/cache_handler.py:139]: This is a comma splice. I suggest changing it to `Error importing module "redis"; it is required for RedisCacheHandler`

Comment (ahesford) [reply spotipy/cache_handler.py]: Done

**Review by stephanebruckert — APPROVED** [review]

**Pull request merged**